ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor.hs

```diff
@@ -667,3 +702,110 @@ wakeupDecision st _now =
     decisionState = st,
     decisionJobs  = []
   }
+
+
+-- | Classify if a node is in promiscuous mode.
+--
+-- A node is not in promiscuous mode only if: it doesn't use ledger peers, peer
+-- sharing, the set of bootstrap peers is empty.
+--
+readAssociationMode
+  :: MonadSTM m
+  => STM m UseLedgerPeers
+  -> PeerSharing
+  -> UseBootstrapPeers
+  -> PeerSelectionSetsWithSizes peeraddr
+  -> STM m AssociationMode
+readAssociationMode
+  readUseLedgerPeers
+  peerSharing
+  useBootstrapPeers
+  PeerSelectionView {
+    viewKnownBootstrapPeers = (knownBootstrapPeers, _)
+  }
+  =
+  do useLedgerPeers <- readUseLedgerPeers
+     pure $
+       case (useLedgerPeers, peerSharing, useBootstrapPeers) of
+         (DontUseLedgerPeers, PeerSharingDisabled, DontUseBootstrapPeers)
+           -> LocalRootsOnly
+         (DontUseLedgerPeers, PeerSharingDisabled, UseBootstrapPeers {})
+           |  Set.null knownBootstrapPeers
```

Comment:
This looks racy to me.
Isn't it possible for us to have no bootstrap peers while we're in the process of doing a DNS lookup on the bootstrap peers?

I think that this check should just be removed.

Reply:
And as soon as we will resolve it, the node will switch to `Unrestricted` mode. Is this a problem?
On the other hand, if we remove the whole case, the property will become static unless one changes the topology file.

Reply:
This covers the case of empty `bootstrap` peers in the topology file with at least one trustable local root peer.  Then `bootstrap` mode will be used ([ref](https://github.com/intersectmbo/cardano-node/blob/388038b8a356aa08e7ff137d255708175036137e/cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs#L306-L320)), and the node will switch to only trustable local roots when it's far from the tip.

We could match on `UseBootstrapPeers [RelayAccessPoint]`, since the list contains unresolved addresses - directly what's written in the topology file.
I think it's a mistake to include `[RelayAccessPoint]` there, but we could have a `UseBootstrapPeers Bool` inside the outbound governor instead.

Reply:
> And as soon as we will resolve it, the node will switch to Unrestricted mode. Is this a problem?

Yes, that is the original case for this hole modification. We where worried that we would signal to consensus that we're in a safe state so fast that we hadn't even had time to connect to an external bootstrap peer.

Changing this be a static config check using `[RelayAccessPoint]` seems fine to me.

Reply:
I am fine with making the property static.  But maybe I misunderstand something, so just for the sake of the argument... If we resolve bootstrap peer and now `readAssociationMode` returns `UnrestrictedMode` instead of `LocalRootsOnly`, we will land in `(Unrestricted, UseBoostrapPeers {})` in `outboundConnectionState`; we will return `TrustedStateWithExternalPeers` only if there's an active bootstrap peer - so how's that too early?

Reply:
oh, I think I got it.  You're worried that we return `LocalRootsOnly` early and then return `TrustedStateWithExternalPeers` if we are connected to a trusted local root, while we should instead wait for an active bootstrap peer.  You're right we actually have to use the configuration value.

Reply:
Fixed.